### DISCUSSION TITLE
Going back history with multiline commands

### DISCUSSION THREAD

**Discussion by romaia** [akinomyoga/ble.sh#440] [Q&A] [answered]
When my last command is a muiltiline comand, and I press up, the last command shows up, with the cursor at the very last of the command.

Now If I press up again, I go to the second last line, pressing up again, the third last line, and so on...

Can I configure some binding that instead of going up one line at a time (in the editor), I actually go back one command at a time?

IE: pressing `C up` shows the last command, pressing it again, shows the command before that (instead of going up one line)?

In other words: quickly go through history, even if there are multiline commands in the history.

**Comment by akinomyoga** [comment, marked as answer]
In the default binding, <kbd>PageUp</kbd> and <kbd>PageDown</kbd> are bound by the history prefix search. If the history prefix search is started *with an empty command line*, it actually moves one history entry for every keypress.

If you would like to configure <kbd>C-up</kbd> and <kbd>C-down</kbd> for the actual history movement, you can use the following settings:

```bash
function my/emacs/set-up-history-movements-with-control {
  # For emacs editing mode
  ble-bind -m emacs -f C-up 'history-prev'
  ble-bind -m emacs -f C-down 'history-next'
}
blehook/eval-after-load keymap_emacs my/emacs/set-up-history-movements-with-control

function my/vi/set-up-history-movements-with-control {
  # For vi editing mode
  ble-bind -m vi_imap -f C-up 'history-prev'
  ble-bind -m vi_imap -f C-down 'history-next'
  ble-bind -m vi_nmap -f C-up 'vi-command/history-prev'
  ble-bind -m vi_nmap -f C-down 'vi-command/history-next'
}
blehook/eval-after-load keymap_vi my/vi/set-up-history-movements-with-control
```

Or you can also configure the same through the readline interface:

```bash
bind -m emacs '"\e[1;5A": previous-history'
bind -m emacs '"\e[1;5B": next-history'
bind -m vi-insert '"\e[1;5A": previous-history'
bind -m vi-insert '"\e[1;5B": next-history'
bind -m vi-command '"\e[1;5A": previous-history'
bind -m vi-command '"\e[1;5B": next-history'
```

**Reply by romaia** [reply]
I usually look for the keyboard with `ble-bind` without any arguments, but I didn't see `PageUp` and `PageDown` there. Should they show up in the list?

**Reply by akinomyoga** [reply]
Yes. <kbd>PageUp</kbd> and <kbd>PageDown</kbd> are represented by `prior` and `next` as listed in a table in [§3.1](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A73-Key-Binding#user-content-sec-kspecs). These names are motivated by those in GNU Emacs.

**Reply by DJCrashdummy** [reply]
as i'm new to ble.sh and just figuring out what works for me and what does not, i was annoyed by the fact to lose the ability to browse through history - no matter how long the commands are - by just one key stroke.

after some research i found this and i'm pretty happy with it, as i guess i can work with `C-up` and `C-down` instead of just `up` and `down` compared to the original bash. and if you think about it a little bit more, it makes even more sense... thank you @akinomyoga for suggesting that!

that leads me to one **question @akinomyoga**: as `C-up` and `C-down` aren't used by default and using them to quickly browse the history seems pretty reasonable, why isn't that set by default?

---

BTW: i created a completely new `$HOME/.blerc` and just inserted 
> ```shell
> # For emacs editing mode
> ble-bind -m emacs -f C-up 'history-prev'
> ble-bind -m emacs -f C-down 'history-next'
> 
> # For vi editing mode
> ble-bind -m vi_imap -f C-up 'history-prev'
> ble-bind -m vi_imap -f C-down 'history-next'
> ble-bind -m vi_nmap -f C-up 'vi-command/history-prev'
> ble-bind -m vi_nmap -f C-down 'vi-command/history-next'
> ```

via copy & paste... but now every time i open a terminal, i get this:
```
ble-bind: Unknown ble edit function `vi-command/history-prev'.
ble-bind: Unknown ble edit function `vi-command/history-next'.
```
as i have no clue about vi or vim, i can't tell if this is expected... for sure everything is fine and still works if i comment resp. remove the vi lines.
i just found it somewhat odd, that 2 of the 4 vi lines produced an error, so i just wanted to mention that.

**Reply by akinomyoga** [reply]
> via copy & paste... but now every time i open a terminal, i get this:
> 
> ```
> ble-bind: Unknown ble edit function `vi-command/history-prev'.
> ble-bind: Unknown ble edit function `vi-command/history-next'.
> ```
> 
> as i have no clue about vi or vim, i can't tell if this is expected...

Ah, OK. As described in [§6.3](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A76-Vim-Mode#user-content-sec-vim-mode-settings), to fully configure the Vim settings, they need to be executed through the "keymap_vi" load hook:


```bash
function my/vim-load-hook {
  # Put your vi editing mode settings here
  ble-bind -m vi_imap -f C-up 'history-prev'
  ble-bind -m vi_imap -f C-down 'history-next'
  ble-bind -m vi_nmap -f C-up 'vi-command/history-prev'
  ble-bind -m vi_nmap -f C-down 'vi-command/history-next'
}
blehook/eval-after-load keymap_vi my/vim-load-hook
```

> i just found it somewhat odd, that 2 of the 4 vi lines produced an error, so i just wanted to mention that.

That is because `history-{prev,next}` are defined in the main `ble.sh`, while `vi-command/history-{prev,next}` is only defined in the [`keymap.vi`](https://github.com/akinomyoga/ble.sh/blob/14f98dd899cd2d95bc37f09264a1b7ed0b99b901/lib/keymap.vi.sh#L4237-L4251) module, whose loading is delayed. The `keymap_vi` load hook is used to run the configurations after loading the `keymap.vi` module.

**Reply by akinomyoga** [reply]
@DJCrashdummy I updated the code examples in my original reply https://github.com/akinomyoga/ble.sh/discussions/440#discussioncomment-9193589. Thank you.

**Reply by DJCrashdummy** [reply]
> That is because history-{prev,next} are defined in the main ble.sh, while vi-command/history-{prev,next} is only defined in the [keymap.vi](https://github.com/akinomyoga/ble.sh/blob/14f98dd899cd2d95bc37f09264a1b7ed0b99b901/lib/keymap.vi.sh#L4237-L4251) module, whose loading is delayed.

that makes sense... thank you for clarifying.

**Comment by akinomyoga** [comment]
@DJCrashdummy Let me create a separate thread because it is a related but separate discussion.

> that leads me to one **question @akinomyoga**: as `C-up` and `C-down` aren't used by default and using them to quickly browse the history seems pretty reasonable, why isn't that set by default?

Hmm, that's actually a good question. I haven't cared about it and haven't noticed that <kbd>C-up</kbd> and <kbd>C-down</kbd> are unused. Come to think of it, yes, I don't have memory that I used <kbd>C-up</kbd> and <kbd>C-down</kbd> in Bash. I think we can bind it to these. I'll think about it.

**Reply by DJCrashdummy** [reply]
if it won't get the default, IMHO this should be added somewhere to the manual, Q&A, wiki or the like as i was pretty surprised to not even find a hint about that.

**Reply by akinomyoga** [reply]
In commit d7afee0054a58cd4376955663d4e97ae6d1f5d4d, I added the default keybindings of <kbd>C-up</kbd> and <kbd>C-down</kbd>.

**Comment by DJCrashdummy** [comment]
though one additional and also somewhat related question @akinomyoga (which would make this feature for me personally perfect):
is it possible to prevent `up` and `down` from browsing the history at all? ...so that no matter what, `C-up`/`C-down` browses the history and `up`/`down` just the current command. this would make working with longer commands really convenient for me.

**Reply by akinomyoga** [reply]
For `emacs` and `vi_imap`, you can configure them by

```bash
# emacs
ble-bind -m 'emacs' -f 'up' 'backward-line'
ble-bind -m 'emacs' -f 'down' 'forward-line'

# vi_imap
ble-bind -m 'vi_imap' -f 'up' 'backward-line'
ble-bind -m 'vi_imap' -f 'down' 'forward-line'
```

It seems there aren't *exactly* corresponding widgets for `vi_nmap`. The [corresponding widgets](https://github.com/akinomyoga/ble.sh/blob/14f98dd899cd2d95bc37f09264a1b7ed0b99b901/lib/keymap.vi.sh#L3262-L3269) seem to hardcode the `history` option. In this case, you can define a modified version:

```bash
# vi_nmap
function ble/widget/my/vi_nmap/forward-line {
  local ARG FLAG REG; ble/keymap:vi/get-arg 1
  ble/widget/vi-command/relative-line.impl "$ARG" "$FLAG" "$REG"
}
function ble/widget/my/vi_nmap/backward-line {
  local ARG FLAG REG; ble/keymap:vi/get-arg 1
  ble/widget/vi-command/relative-line.impl "$((-ARG))" "$FLAG" "$REG"
}
ble-bind -m 'vi_nmap' -f 'up' 'my/vi_nmap/backward-line'
ble-bind -m 'vi_nmap' -f 'down' 'my/vi_nmap/forward-line'
```

I haven't actually tested the above settings, so could you confirm they work?

**Reply by DJCrashdummy** [reply]
thanks @akinomyoga for the quick reply!
i didn't think it was that easy, as the default binding for both `up` and `down` has just a closing `history` in it (`ble-bind -m 'emacs' -f 'down' '... history'`) instead of `history-prev` resp. `history-next`. ...so i assumed that there is something hardcoded.

---

as i never used vi/vim, i have no clue how to work with it... i don't even know how i get form the "insert mode" to the "normal mode". :shrug:
so unfortunately i can just confirm that in the emacs mode and the default vi insert mode (vi_imap) everything works as expected. :+1:

**Reply by akinomyoga** [reply]
If you don't use the vi editing mode and don't know how to test it, that's fine. Thank you for checking the behavior with the emacs editing mode!

FYI, you can switch to the normal mode by pressing <kbd>Escape</kbd>, <kbd>C-[</kbd>, or <kbd>C-c</kbd> in the insert mode. If you want to come back to the insert mode, you can press <kbd>i</kbd>, <kbd>a</kbd>, or many other different commands with additional effects.
